secrets/provider_registry.go

```diff
@@ -0,0 +1,131 @@
+package secrets
+
+import (
+	"encoding/json"
+	"fmt"
+	"sync"
+
+	"github.com/buildkite/agent/v3/bootstrap/shell"
+	"github.com/puzpuzpuz/xsync"
+)
+
+type ProviderRegistry struct {
+	config ProviderRegistryConfig
+	// We have the two maps here because we only want to initialise a provider if a secret using that provider is used
+	// We shouldn't boot up any secrets providers that won't get used
+	candidates *xsync.MapOf[string, providerCandidate] // Candidates technically doesn't need to be a sync map as it's never altered after initialization, but i've made it one just for symmetry
```

Comment:
generics! 🎉

Reply:
more external libraries 👎

Reply:
The question will be: is there a significant benefit to using the library over either the inbuilt map with a sync.Mutex/sync.RWMutex, or using sync.Map (are there going to be that many concurrent map reads/writes that the performance is noticeably better?)

Reply:
yeah, this was originally implemented with a normal map and a `sync.RWMutex`, but there's a lot of reads and writes that require opening and closing the mutex a whole bunch, and the code got really ugly.

then i tried a `sync.Map`, but it's not typesafe, and casting everything gets really ugly...

my (completely unresearched) vibe on the number of secrets per pipeline is P95<10, P99<100, so i feel like we're unlikely to see a level of concurrency where we'd see a lot of lock contention, regardless of implementation.

the other thing is that i don't think my locking dynamics here are necessarily correct; i think that two secrets could initialise the same provider at the same time, but then only one of them would get stored, which isn't fab. i think this code needs more work regardless of what locking system we use.

TL;DR: i was experimenting a bit. i'm happy to make these stdlib `sync.Map`s, or a regular map behind a mutex, but i do kinda like having the fancy typesafe concurrent map.